# Cleaning Data 
take a critical look at your dataset to fix any issues before you move on to a more sophisticated analysis.

In [1]:
import pandas as pd

In [6]:
nba = pd.read_csv("Desktop/nba_all_elo.csv")

In [7]:
nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  ...   win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...   40.294830     NYK    Knicks       68   
1            0     NYK   Knicks  ...   41.705170     TRH   Huskies       66   
2            0     CHS    Stags  ...   42.012257     NYK    Knicks       47   
3            0     NYK   Knicks  ...   40.692783     CHS     Stags       63   
4            0     DTF  Falcons  ...   38.864048     WSC  Capitols       50   

   opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0  1300.0000  1306.7233              H            L  0.640065   NaN  
1  1300.0000  1293.2767              A            W  0.359935   NaN  
2  1306.7233  1297.0712              H            W  0.631101   NaN  
3  1300.0000  1309.6521              A            L  0.368899   NaN  
4  1300.0000  1320.3811              H            L  0.640065   NaN  

[5 rows x 23 columns]

# Missing Values
Have you ever wondered why .info() shows how many non-null values a column contains? The reason why is that this is vital information. Null values often indicate a problem in the data-gathering process. They can make several analysis techniques, like different types of machine learning, difficult or even impossible.

When you inspect the nba dataset with nba.info(), you’ll see that it’s quite neat. Only the column notes contains null values for the majority of its rows:

In [8]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
gameorder        126314 non-null int64
game_id          126314 non-null object
lg_id            126314 non-null object
_iscopy          126314 non-null int64
year_id          126314 non-null int64
date_game        126314 non-null object
seasongame       126314 non-null int64
is_playoffs      126314 non-null int64
team_id          126314 non-null object
fran_id          126314 non-null object
pts              126314 non-null int64
elo_i            126314 non-null float64
elo_n            126314 non-null float64
win_equiv        126314 non-null float64
opp_id           126314 non-null object
opp_fran         126314 non-null object
opp_pts          126314 non-null int64
opp_elo_i        126314 non-null float64
opp_elo_n        126314 non-null float64
game_location    126314 non-null object
game_result      126314 non-null object
forecast         126314 non-null float64
notes     

This output shows that the notes column has only 5424 non-null values. That means that over 120,000 rows of your dataset have null values in this column.

# ignore missing values

Sometimes, the easiest way to deal with records containing missing values is to ignore them. You can remove all the rows with missing values using .dropna():

In [12]:
rows_without_missing_data = nba.dropna(axis=1)

In [13]:
rows_without_missing_data.shape

(126314, 22)

Now, the resulting DataFrame contains all 126,314 games, but not the sometimes empty notes column.

***

# replace the missing values
If there’s a meaningful default value for your use case, then you can also replace the missing values with that:

In [14]:
data_with_default_notes = nba.copy()
data_with_default_notes['notes'].fillna(
    value  = 'no data',
    inplace = True
)

In [19]:
data_with_default_notes['notes'].describe()

count      126314
unique        232
top       no data
freq       120890
Name: notes, dtype: object

Here, empty notes rows is filled with the string "no data"

***

# Invalid Values
Invalid values can be even more dangerous than missing values. Often, you can perform your data analysis as expected, but the results you get are peculiar. This is especially important if your dataset is enormous or used manual entry. Invalid values are often more challenging to detect, but you can implement some sanity checks with queries and aggregations.

One thing you can do is validate the ranges of your data. For this, .describe() is quite handy.

In [20]:
nba.describe()

gameorder        _iscopy        year_id     seasongame  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean    31579.000000       0.500000    1988.200374      43.533733   
std     18231.927643       0.500002      17.582309      25.375178   
min         1.000000       0.000000    1947.000000       1.000000   
25%     15790.000000       0.000000    1975.000000      22.000000   
50%     31579.000000       0.500000    1990.000000      43.000000   
75%     47368.000000       1.000000    2003.000000      65.000000   
max     63157.000000       1.000000    2015.000000     108.000000   

         is_playoffs            pts          elo_i          elo_n  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean        0.063857     102.729982    1495.236055    1495.236055   
std         0.244499      14.814845     112.139945     112.461687   
min         0.000000       0.000000    1091.644500    1085.774400   
25%         0.000000      93.000000    1417.237975    1416.994900   
50%         0.000000     103.000000    1500.945550    1500.954400   
75%         0.000000     112.000000    1576.060000    1576.291625   
max         1.000000     186.000000    1853.104500    1853.104500   

           win_equiv        opp_pts      opp_elo_i      opp_elo_n  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean       41.707889     102.729982    1495.236055    1495.236055   
std        10.627332      14.814845     112.139945     112.461687   
min        10.152501       0.000000    1091.644500    1085.774400   
25%        34.103035      93.000000    1417.237975    1416.994900   
50%        42.113357     103.000000    1500.945550    1500.954400   
75%        49.635328     112.000000    1576.060000    1576.291625   
max        71.112038     186.000000    1853.104500    1853.104500   

            forecast  
count  126314.000000  
mean        0.500000  
std         0.215252  
min         0.020447  
25%         0.327989  
50%         0.500000  
75%         0.672011  
max         0.979553

# Year_id
The year_id varies between 1947 and 2015. That sounds plausible.
# pts
What about pts? How can the minimum be 0? Let’s have a look at those games:

In [21]:
nba[nba['pts']==0]

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
26684      13343  197210260VIR   ABA        1     1973  10/26/1972   

       seasongame  is_playoffs team_id  fran_id  \
26684           7            0     DNR  Nuggets   

                    ...                win_equiv  opp_id  opp_fran  opp_pts  \
26684               ...                40.408863     VIR   Squires        2   

       opp_elo_i opp_elo_n  game_location  game_result  forecast  \
26684  1484.1907  1487.083              A            L  0.328948   

                                notes  
26684  at Richmond VA; forfeit to VIR  

[1 rows x 23 columns]

It seems the game was left/surrendered. Depending on your analysis, you may want to remove it from the dataset.

# Inconsistent Values
Sometimes a value would be entirely realistic in and of itself, but it doesn’t fit with the values in the other columns. You can define some query criteria that are mutually exclusive and verify that these don’t occur together.

In the NBA dataset, the values of the fields pts, opp_pts and game_result should be consistent with each other. You can check this using the .empty attribute:

In [22]:
nba[(nba["pts"] > nba["opp_pts"]) & (nba["game_result"] != 'W')].empty

True

In [23]:
nba[(nba["pts"] > nba["opp_pts"]) & (nba["game_result"] != 'L')].empty

False

In [24]:
nba[(nba["pts"] < nba["opp_pts"]) & (nba["game_result"] != 'W')].empty

False

In [25]:
nba[(nba["pts"] < nba["opp_pts"]) & (nba["game_result"] != 'L')].empty

True